cirq-core/cirq/_compat.py

```diff
@@ -27,21 +27,24 @@
 import numpy as np
 import pandas as pd
 import sympy
+import sympy.printing.repr
 
 
 def proper_repr(value: Any) -> str:
     """Overrides sympy and numpy returning repr strings that don't parse."""
 
     if isinstance(value, sympy.Basic):
-        result = sympy.srepr(value)
-
         # HACK: work around https://github.com/sympy/sympy/issues/16074
-        # (only handles a few cases)
-        fixed_tokens = ['Symbol', 'pi', 'Mul', 'Pow', 'Add', 'Mod', 'Integer', 'Float', 'Rational']
-        for token in fixed_tokens:
-            result = result.replace(token, 'sympy.' + token)
+        fixed_tokens = dir(sympy)
+
+        class Printer(sympy.printing.repr.ReprPrinter):
+            def _print(self, expr, **kwargs):
+                s = super()._print(expr, **kwargs)
+                if any(s.startswith(t) for t in fixed_tokens):
+                    return 'sympy.' + s
+                return s
 
-        return result
+        return Printer().doprint(value)
```

Comment:
It looks like this might be possible using `sympy.srepr(value)`?

Also, this feels separate from the rest of the PR, and is pretty wide-reaching - I'd prefer to isolate it in its own PR instead of burying it here.

Reply:
That PR one would have to go in first, since we need to serialize all the sympy comparison operators now. I added some additional sympy serialization tests in the latest commit. Do you want to pull all those into a separate PR?

Reply:
Created #4767. There I limited the number of classes it looks for, rather than all of dir(sympy).

No, it's not possible to use `sympy.srepr(value)`, as that's what was there prior to this PR. If we just add the relational operators to the existing implementation, then things get munged when it substitutes "LessThan" and "StrictLessThan" for instance. (I'm surprised we haven't seen that already when it replaces "pi" in the middle of some word). So here we override the ReprPrinter class and make sure that we're only replacing the beginnings of strings. Still not perfect, but better than it was.

Reply:
Ah, literally the bug the link points to 😝 #4767 looks like what I had in mind, thanks!